### PR TITLE
DB Restructure

### LINKED ISSUE #45: update db structure
New DB structure. 

**Overview:** A `Contest` is the umbrella container for everything. Each `Contest` has one `Waiting Room` that has sign up dates tied to it that can be set and edited by an admin as needed. `Waiting Rooms` get split into `Competitions`. Competitions hold groups of users competing in a contest and store the date the competition starts and the date they will end based on the run period of the `Contest`.

#### Contests
* `contest_id`
* `campaign_id` 
* `campaign_run_id`
* `waiting_room_id`
* How long the contest runs for (`duration`)

Also has (currently undetermined) metadata such as: 
* title 
* description
* prize (?)
* rules

#### Waiting Rooms
* `waiting_room_id`
* `contest_id`
* `signup_start_date` (admin can edit)
* `signup_end_date` (admin can edit)

#### Waiting Room Users
NOTE: I think this table exists already and should still work with this new structure:
`waiting_room_id`
`user_id`

#### Competitions
* `competition_id`
* `contest_id`
* `competition_start_date` (based on the day this competition was created)
* `competition_end_date` (calculated based on `run_period` and set on creation

#### Competition Users
* `competition_id`
* `user_id`

### PR DESCRIPTION
**Pull request by sbsmith86**
#### What's this PR do?

Updates the DB structure in Gladiator as outlined in #45 

* Removed old migrations, and made new ones. 
* Added a new `Contests` table and a model that defined it's relationship to waiting rooms and competitions
* Renamed pivot tables so they get run last. 
* Added a new `ContestTableSeeder` that makes 3 contests.
* Added a `ContestServiceProvider` that listens for the `created` event on Contest models and automatically creates a new waiting room for the competition. 
* Created a competition seeder that just creates three empty competitions for a Contest.

NOTES/QUESTIONS -

* I wasn't sure about the `duration` field. My thought is we can make the decision to always store this as an `integer` representing days and then on the front end we can provide a slicker interface with like units (days, weeks, months, etc) that gets converted into days before it gets stored in the db. But I am open to doing something different. 

* Because of the ^^ open question I didn't implement any fancy calculation of competition start and end date yet. 

* I didn't touch any CRUD stuff or other logic to keep this as small as I could, so we can fix those up in other issues. 

#### How should this be manually tested?

Pull down this branch
Run `composer dump-autoload`
Run `php artisan:refresh` to get new tables
Run `php artisan db:seed` to seed the DB

Review the DB to make sure all the correct columns are there and the relationships look right.

#### What are the relevant tickets?

Fixes #45 

@angaither @deadlybutter @weerd

### CODE REVIEW

**Comment by angaither** [comment]
my big feeling on `duration` on contests is that it probably shouldn't be there. 
we have dates on the waiting room and that should be accessible to the competition but shouldn't be a different value since it's just so many dates.

**Comment by sbsmith86** [comment]
@angaither The waiting room dates designate the length of the sign up period and the competition duration is supposed to help us track how long a competition will run for (so like how long each group of people will actually have to compete) How would we do that without `duration`?

**Comment by angaither** [comment]
@sbsmith86 ahh, my bad. I thought `duration` on the `contest` was how long the signup period should be open. totally misunderstood that field

File: app/Providers/ContestServiceProvider.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        Contest::created(function ($contest) {
+            $waitingRoom = new WaitingRoom;
+
+            $waitingRoom->contest_id = $contest->getKey();
+
+            // Default sign up period of one week.
+            $waitingRoom->signup_start_date = Carbon::now();
+            $waitingRoom->signup_end_date = Carbon::now()->addWeeks(1)->endOfDay();
+
+            $waitingRoom->save();
```
Comment (weerd) [inline app/Providers/ContestServiceProvider.php]: Nice!

**Comment by weerd** [comment]
> @angaither The waiting room dates designate the length of the sign up period and the competition duration is supposed to help us track how long a competition will run for (so like how long each group of people will actually have to compete) How would we do that without duration?

@sbsmith86 Is it the competition `duration` or the contest `duration`?

**Comment by sbsmith86** [comment]
@weerd ah yea sorry. it's the Contest `duration`...new terminology got me so confused lol. I'm really working on it. One step at a time.

**Comment by angaither** [comment]
now I'm confused again,
i thought it was the `competition` duration that's the first piece of metadata on the contest that's inherited on the all the `competitions`

**Comment by sbsmith86** [comment]
@weerd @angaither it is the `duration` field on the `contests` table that will be used to set `competition_start_date` and `competition_end_date` for each competition in a contest. This way, competitions can start at different times (based on when they were created) but all run for the same amount of time.

**Comment by weerd** [comment]
No worries. So yeah, if it's the contest `duration`, I'm still not sure why we couldn't have dates for that: `contest_start` & `contest_end`, which otherwise match the campaign. A `duration` could easily be calculated from the start and end dates. I think I'm confused too though.

If it's the competition `duration` then that would dictate how long a single competition in the contest has to run, with each competition in a contest starting at different points; but they all run for say a duration of 10 days...

**Comment by weerd** [comment]
Ok, so maybe what's confusing is this comment:

> `$table->integer('duration')->comment = 'Duration of the campaign in days';`

campaign -> competition?

**Comment by weerd** [comment]
> @weerd @angaither it is the duration field on the contests table that will be used to set competition_start_date and competition_end_date for each competition in a contest. This way, competitions can start at different times (based on when they were created) but all run for the same amount of time.

This makes sense; just change up the comment added to the `duration` column in the migration ;)

**Comment by sbsmith86** [comment]
Sorry for the confusion. 

* So an admin creates a contest and says each competition within that contest should run for 1 month (`duration`)
* A waiting room is created and the admin can edit the start and end dates of the waiting room, so lets say they want to collect signups for two weeks. 
* Two weeks go by and the admin is ready to break people out into competitions. 
* They split the waiting room
* The competitions are created and gladiator uses the `duration` to determine the start and end date for each competition group. 
* BUT based on the amount of people, we only split into two competitions and the admin wants to do more signups. 
* they update the waiting room period to new dates
* that signup period is over, and they split the waiting room
* the `duration` field is then used again to figure out the start and end dates for the competitions that were just created.

**Comment by deadlybutter** [comment]
^ what @sbsmith86 just said, :+1:

**Comment by sbsmith86** [comment]
@weerd My bad! yea that comment is :hankey:

**Comment by weerd** [comment]
@sbsmith86 excellent example scenario! very helpful! 

looks good :+1:

**Comment by angaither** [comment]
:+1: 
nice job on this.

**Pull request merged**